transmart-core-api/src/main/groovy/org/transmartproject/core/tree/TreeNode.groovy

```diff
@@ -0,0 +1,128 @@
+package org.transmartproject.core.tree
+
+import org.transmartproject.core.ontology.OntologyTerm
+import org.transmartproject.core.ontology.OntologyTermTag
+import org.transmartproject.core.ontology.OntologyTermType
+
+/**
+ * The ontology node class that represents nodes in the ontology forest.
+ * (There may be multiple root nodes, hence the name 'forest'.)
+ * Intermediate nodes contain a collection of child nodes.
+ * Leaf nodes represent a concept (either variable or value) or a study in TranSMART.
+ */
+interface TreeNode {
+
+    /**
+     * API version string used for marshalling the object.
+     * @return the API version string.
+     */
+    String getApiVersion()
+
+    void setApiVersion(String version)
+
+    /**
+     * The domain object on which this node is based.
+     */
+    OntologyTerm getDelegate()
+
+    /**
+     * Retrieves the parent node if it exists (i.e., when it is not a root node).
+     * @return the parent node.
+     */
+    TreeNode getParent()
+
+    void setParent(TreeNode parent)
+
+    /**
+     * Retrieves the child nodes.
+     * @return the child nodes.
+     */
+    List<TreeNode> getChildren()
+
+    /**
+     * Retrieves the tags associated with the node.
+     * @return the list of tags.
+     */
+    List<OntologyTermTag> getTags()
+
+    void setTags(List<OntologyTermTag> tags)
+
+    /**
+     * The number of observations that satisfy the constraint in {@link #getConstraint()}.
+     * The count is not always set, it can also be null.
+     * @return the number of observations for this node if set; null otherwise.
+     */
+    Long getObservationCount()
+
+    void setObservationCount(Long count)
+
+    /**
+     * The number of patients that have observations that satisfy the constraint in {@link #getConstraint()}.
+     * The count is not always set, it can also be null.
+     * @return the number of patients for this node if set; null otherwise.
+     */
+    Long getPatientCount()
+
+    void setPatientCount(Long count)
+
+    /**
+     * The name of the node.
+     * @return the name.
+     */
+    String getName()
+
+    /**
+     * The full tree path that contains all ancestors from the root up to the current node and the name of the node.
+     * @return the path.
+     */
+    String getFullName()
+
+    /**
+     * The visual attributes that characterise how the node should be represented.
+     * @return the set of visual attributes.
+     */
+    EnumSet<OntologyTerm.VisualAttributes> getVisualAttributes()
+
+    /**
+     * The dimension that the current node refers to. Could be study or concept.
+     */
+    String getDimension()
```

Comment:
Why we repeat ontology term field in here? We could always access them via `delegate` field. BTW maybe we should come up with a better name in the core api for this field.

Reply:
In this case, because we need to be able to retrieve the tags for a tree node and the tagsResource contract is based on OntologyTerm objects.
We could change that one of course, to create a `Map<String, List<OntologyTermTag>> getTags(Set<String> ontologyTermPaths, boolean includeDescendantsTags)`. I will look into it.

Reply:
Fixed.